src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -2513,47 +2513,86 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
     size_t   op1Val   = (size_t)trueVal->AsIntCon()->IconValue();
     size_t   op2Val   = (size_t)falseVal->AsIntCon()->IconValue();
 
-    if (op1Val + 1 == op2Val || op2Val + 1 == op1Val)
+    if ((op1Val + 1 == op2Val) || (op2Val + 1 == op1Val))
     {
-        const bool shouldReverseCondition = op1Val + 1 == op2Val;
+        LowerToCincOrCinv(select, cond, (op1Val + 1 == op2Val), true);
+    }
+}
+
+//----------------------------------------------------------------------------------------------
+// TryLowerCselToCinv: Try converting SELECT/SELECTCC to CINV/CINVCC. Conversion is possible only if the
+// trueVal == ~(falseVal).
+//
+// Arguments:
+//     select - The select node that is now SELECT or SELECTCC
+//     cond   - The condition node that SELECT or SELECTCC uses
+//
+void Lowering::TryLowerCselToCinv(GenTreeOp* select, GenTree* cond)
+{
+    assert(select->OperIs(GT_SELECT, GT_SELECTCC));
+
+    GenTree* trueVal       = select->gtOp1;
+    GenTree* falseVal      = select->gtOp2;
+    GenTree* negatedVal    = (trueVal->gtOper == GT_NOT) ? trueVal : falseVal;
+    GenTree* nonNegatedVal = (trueVal->gtOper == GT_NOT) ? falseVal : trueVal;
 
-        // Create a cinc node, insert it and update the use.
-        if (select->OperIs(GT_SELECT))
+    if (GenTree::Compare(negatedVal->AsOp()->gtOp1, nonNegatedVal))
```

Comment:
`GenTree::Compare` is a syntactic check, it is not sufficient for what you want to check here. For example, both operands could be syntactically similar calls, but that doesn't mean they both compute to the same value. They could also be e.g. two `LCL_FLD` nodes with an interfering assignment between them.
`TryLowerConditionToFlagsNode` does a similar check to transform FP `x != x` for locals on xarch that you could look at for a correct way of doing this.

Reply:
I have [updated](https://github.com/dotnet/runtime/pull/84926/files#diff-547b75c8f8dd006693805c30d7a0e2ddcdd6b10691383dfbb16e3ed0c50f236aR2539) the patch to do the `IsInvariantInRange` check, similar to `x != x`. However. I don't fully understand it. Could you take a look, please?

Reply:
I think you should scope the check down to just locals, like the other check I referenced -- the call example I pointed out would otherwise still cause problems.

Reply:
Sure. Scoping down to locals is missing the cases where the local variables are being casted. e.g., in the case of bytes the local `int` are cast to `ubyte`.  Consequently, we are getting missing an opportunity to use `cinv`.

In the following example
```
static byte cinv_byte(byte op1, byte op2)
{
     return (byte) (op1 > 42 ? op2: ~op2);
}
```

The `t17` and `t18` are the casts.

```
N001 (  2,  2) [000000] -----------                    t0 =    LCL_VAR   int    V00 arg0         u:1 (last use) $80
                                                            /--*  t0     int
N002 (  3,  4) [000016] -----------                   t16 = *  CAST      int <- ubyte <- int $100
N003 (  1,  2) [000001] -c---------                    t1 =    CNS_INT   int    42 $44
                                                            /--*  t16    int
                                                            +--*  t1     int
N004 (  5,  7) [000002] J------N---                    t2 = *  GT        int    $101
N005 (  2,  2) [000004] -----------                    t4 =    LCL_VAR   int    V01 arg1         u:1 (last use) $81
                                                            /--*  t4     int
N006 (  3,  4) [000018] -----------                   t18 = *  CAST      int <- ubyte <- int $102
N007 (  2,  2) [000011] -----------                   t11 =    LCL_VAR   int    V01 arg1         u:1 (last use) $81
                                                            /--*  t11    int
N008 (  3,  4) [000017] -----------                   t17 = *  CAST      int <- ubyte <- int $102
                                                            /--*  t17    int
N009 (  4,  5) [000012] -----------                   t12 = *  NOT       int    $140
                                                            /--*  t2     int
                                                            +--*  t18    int
                                                            +--*  t12    int
N010 ( 13, 17) [000024] -----------                   t24 = *  SELECT    int
```

Missing this optimisation didn't change the asmdiffs. It may indicate that this is a relatively less common pattern and thus can be skipped. I'll update the patch scoped down to locals. Let me know if you would like to incorporate cinv for bytes and shorts.